### ISSUE TITLE
Option List from API

### ISSUE THREAD

**Issue by mcockrellsana** [cofoundry-cms/cofoundry#320]
Hey Joel,

I'm encountering odd issues when using IListOptionApiSource.  I've stepped through the code for the CmsFormFieldDropDown and its calling my api, its returning my values ( the return type is a json array like this [{ name: "somename", abbreviation: "sn" }, { name: "othername", abbreviation: "on" }] ).  But what I see is that it isnt binding my values, I get about 5 or 6 undefined options in the drop down.  The option name and option value accessors are name and abbreviation so they should match, yet they dont.  

The weird thing is when I look at the return from the $http.get call in the component, its wrapped with other stuff, and my json array is sitting in a data property.  I've been fiddling with this for a couple hours,  seems so simple yet I cant figure out why my simple json array won't bind to the drop down.

Thanks for all your help.

**Comment by HeyJoel** [comment]
It seems that the $http call is returning the raw request instead of the data from the response. This might be a regression and I'll have to do some investigation as to why this behaviour has changed. I should have some time today to look at it in more depth.

**Comment by HeyJoel** [comment]
Ok I found the problem, it is expecting the API to be located in the admin panel directory. HTTP requests from this directory are handled differently than those from external resources, including response formatting and error handling. If located outside of the admin path then the $http response is not intercepted at all.

### Workaround

Locate your api in the admin panel directory, which by default is *'/admin'* e.g. *'/admin/api/pets'*

### Fix

To fix this it would probably be good to run these API calls through a service that specializes in parsing the response correctly for this specific use case. 

The problem really stems from using $http as a singleton and using http interceptors to modify the response, whereas it's probably a better idea to wrap $http in a new service so you can decide which service to use depending on the api call you want to make and how you want to handle errors.

**Comment by mcockrellsana** [comment]
Your workaround and fix make perfect sense, thank you for the issue investigation.

**Comment by TiriTM** [comment]
I currently have (at least) a similar issue, the response from the api call is a json array of objects with the properties referenced as `NameField` and `ValueField` in the `ProductOptionSource` class.

```
public class ProductOptionSource : IListOptionApiSource
{
	public string Path => "/admin/api/content/products";
	public string NameField => "distinctiveTitle";
	public string ValueField => "productReference";
}
```

But the options in the dropdown are only displayed as `undefined` and their values as `undefined:undefined`.

As you can see the API called is already located in the admin panel directory and its response is actually (anonymized because of company's secrets) `[{productReference: "978xxxxxxxxx8", ..., distinctiveTitle: "Lorem ipsum dolor sit amet consectetuer adipiscing elit", ... }, {...}, {...}]` what's wrong with this?

Kind regards.

**Comment by HeyJoel** [comment]
@TiriTM sorry the documentation isn't explicit, but you need to wrap your API response in a data object. The simplest way to make sure it's in the right format is to use our `IApiResponseHelper`:

```csharp
public class PetsController : Controller
{
    private readonly IApiResponseHelper _apiResponseHelper;

    public PetsController(
        IApiResponseHelper apiResponseHelper
        )
    {
        _apiResponseHelper = apiResponseHelper;
    }

    [Route("admin/api/pets")]
    public IActionResult Pets()
    {
        var results = new object[] {
            new { Id = 1, Title = "Dog" },
            new { Id = 2, Title = "Cat" },
            new { Id = 3, Title = "Fish" },
            new { Id = 4, Title = "Stick Insect" }
        };

        return _apiResponseHelper.SimpleQueryResponse(this,results);
    }
}
````
So the response looks like this:

```json
{
  "data": [
    {
      "id": 1,
      "title": "Dog"
    },
    {
      "id": 2,
      "title": "Cat"
    },
    {
      "id": 3,
      "title": "Fish"
    },
    {
      "id": 4,
      "title": "Stick Insect"
    }
  ]
}
```

I think I might have made this optional in the fix for this issue, but I'll make a note of it here and check it over when I have time to look at the code next.

**Comment by TiriTM** [comment]
@HeyJoel, thanks for your reply, unfortunately it doesn't work right up to now. We're using an async Task<IActionResult> and originally our CMSAPIController derives from ControllerBase.

```
[Route("admin/api/content")]
[ApiController]
public class CMSAPIController : Controller // ControllerBase led to an error with ApiResponseHelper
{
	private readonly IAsyncWebProductDAL _products;
	private readonly IApiResponseHelper _apiResponseHelper;

	public CMSAPIController(IApiResponseHelper apiResponseHelper, IAsyncWebProductDAL products)
	{
		_apiResponseHelper = apiResponseHelper;
		_products = products;
	}

	[Route("products")]
	public async Task<IActionResult> GetProducts(String q = null)
	{
		if (q == null)
		{
			q = string.Empty;
		}

		var resultData = await _products.FindAsync(q, 1, 10, false, true).ConfigureAwait(false);
		var response = _apiResponseHelper.SimpleQueryResponse(this, resultData.Results);
		return Ok(response);
	}
}
```

The object returned by `_apiResponseHelper.SimpleQueryResponse(this, resultData.Results)` - and stored in `response` for debugging reasons - looks like this:

```
{
	contentTypes: [],
	declaredType: null,
	formatters: [],
	statusCode: 200,
	value: {
		data: [,...]
	}
}
```

Actually our API-Response **is** inside of the data array, but the dropdown still shows "undefined". Did I miss anythintg?

**Comment by HeyJoel** [comment]
It shouldn't matter if the method is async, the problem is that ` _apiResponseHelper.SimpleQueryResponse` returns `IActionResult` which you are wrapping in another `IActionResult` by calling `OK(result)`. Just return the result directly:

```csharp
[Route("admin/api/content")]
[ApiController]
public class CMSAPIController : Controller
{
	private readonly IAsyncWebProductDAL _products;
	private readonly IApiResponseHelper _apiResponseHelper;

	public CMSAPIController(IApiResponseHelper apiResponseHelper, IAsyncWebProductDAL products)
	{
		_apiResponseHelper = apiResponseHelper;
		_products = products;
	}

	[Route("products")]
	public async Task<IActionResult> GetProducts(String q = null)
	{
		if (q == null)
		{
			q = string.Empty;
		}

		var resultData = await _products.FindAsync(q, 1, 10, false, true);
		var response = _apiResponseHelper.SimpleQueryResponse(this, resultData.Results);
		return response;
	}
}
```

Also [You don't need `ConfigureAwait` in ASP.NET Core](https://blog.stephencleary.com/2017/03/aspnetcore-synchronization-context.html).

**Comment by TiriTM** [comment]
@HeyJoel okay that absolutely makes sense and it solved the issue 🥇

**Comment by HeyJoel** [comment]
Fixed, will be release in 0.9.:

- APIs haven't need to be in the /admin directory for a while, but I've now updated the docs
- APIs can return data either in the document root, or in a `data` wrapper object.
